### ISSUE TITLE
json_clean doesn't know how to serialize pandas/numpy types

### ISSUE THREAD

**Issue by jtyberg** [ipython/ipykernel#84]
I'm using the [jupyter-incubator/declarativewidgets](https://github.com/jupyter-incubator/declarativewidgets) in a notebook in 4.x.  When I try to send a dictionary that I created from a pandas Series (to_dict()), I'm running into a JSON serialization error.  Apparently, sending numpy data types over the widget channels causes the error:

```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-14-329f8d502429> in <module>()
----> 1 channel('pickup').set('pickup', pickup.to_dict())

/opt/conda/lib/python3.4/site-packages/urth/widgets/widget_channels.py in set(self, key, value, **kwargs)
     68     def set(self, key, value, **kwargs):
     69         global the_channels
---> 70         the_channels.set(key, value, self.chan, **kwargs)
     71 
     72     def watch(self, key, handler):

/opt/conda/lib/python3.4/site-packages/urth/widgets/widget_channels.py in set(self, key, value, chan, **kwargs)
     37         attr = "{}:{}".format(chan, key)
     38         serialized = self.serializer.serialize(value, **kwargs)
---> 39         self._send_update(attr, serialized)
     40 
     41     def watch(self, key, handler, chan='default'):

/opt/conda/lib/python3.4/site-packages/urth/widgets/urth_widget.py in _send_update(self, attribute, value)
     29             }
     30         }
---> 31         self._send(msg)
     32 
     33     def send_status(self, status, msg=""):

/opt/conda/lib/python3.4/site-packages/ipywidgets/widgets/widget.py in _send(self, msg, buffers)
    449     def _send(self, msg, buffers=None):
    450         """Sends a message to the model in the front-end."""
--> 451         self.comm.send(data=msg, buffers=buffers)
    452 
    453 

/opt/conda/lib/python3.4/site-packages/ipykernel/comm/comm.py in send(self, data, metadata, buffers)
    125         """Send a message to the frontend-side version of this comm"""
    126         self._publish_msg('comm_msg',
--> 127             data=data, metadata=metadata, buffers=buffers,
    128         )
    129 

/opt/conda/lib/python3.4/site-packages/ipykernel/comm/comm.py in _publish_msg(self, msg_type, data, metadata, buffers, **keys)
     73         data = {} if data is None else data
     74         metadata = {} if metadata is None else metadata
---> 75         content = json_clean(dict(data=data, comm_id=self.comm_id, **keys))
     76         self.session.send(self.iopub_socket, msg_type,
     77             content,

/opt/conda/lib/python3.4/site-packages/ipykernel/jsonutil.py in json_clean(obj)
    161         out = {}
    162         for k,v in iteritems(obj):
--> 163             out[unicode_type(k)] = json_clean(v)
    164         return out
    165     if isinstance(obj, datetime):

/opt/conda/lib/python3.4/site-packages/ipykernel/jsonutil.py in json_clean(obj)
    161         out = {}
    162         for k,v in iteritems(obj):
--> 163             out[unicode_type(k)] = json_clean(v)
    164         return out
    165     if isinstance(obj, datetime):

/opt/conda/lib/python3.4/site-packages/ipykernel/jsonutil.py in json_clean(obj)
    161         out = {}
    162         for k,v in iteritems(obj):
--> 163             out[unicode_type(k)] = json_clean(v)
    164         return out
    165     if isinstance(obj, datetime):

/opt/conda/lib/python3.4/site-packages/ipykernel/jsonutil.py in json_clean(obj)
    161         out = {}
    162         for k,v in iteritems(obj):
--> 163             out[unicode_type(k)] = json_clean(v)
    164         return out
    165     if isinstance(obj, datetime):

/opt/conda/lib/python3.4/site-packages/ipykernel/jsonutil.py in json_clean(obj)
    167 
    168     # we don't understand it, it's probably an unserializable object
--> 169     raise ValueError("Can't clean for JSON: %r" % obj)

ValueError: Can't clean for JSON: 1
```

Here's a notebook that reproduces the error condition: https://gist.github.com/jtyberg/63fe41605566fc1fe866

**Comment by jtyberg** [comment]
I just bumped to pandas 0.17.1 on ipython 4.0.0 and ipykernel 4.1.1, and I'm seeing more of the JSON serialization problems, this time with pandas dates and timedeltas.  See gist: https://gist.github.com/jtyberg/63fe41605566fc1fe866

**Comment by takluyver** [comment]
I don't think json_clean is meant to coerce arbitrary Python data into JSON. Comms send JSON, and I think it's up to the code using them to produce arbitrary JSONable data.

`json_clean` is mainly intended for internal use by IPython. I'm not even sure that we should be calling it on comm data.

**Comment by minrk** [comment]
Raising errors on unrecognised types is the point of json_clean. You need to use custom serialization if you want to send arbitrary objects that aren't basic JSON types. @SylvainCorlay has some examples of widgets sending numpy arrays.

**Comment by jtyberg** [comment]
OK.  I'll try to find those examples.  I see it handles datetime objects.  What about datetime.date?  Where does it stop?

**Comment by minrk** [comment]
It stops at datetime for use in some message headers, and probably shouldn't even be used for that. It's mostly there to iron out some weirdness with Python 2 encodings and other edge cases. In general, you shouldn't pass anything to a comm message for which `json.dumps(obj)` fails. Widgets offer a `to_json` method for custom serialization of data.

**Comment by jtyberg** [comment]
I buy that.  This only started happening when I moved to 4.x.  I was sending pandas/numpy types just fine on 3.2.x, so there must have been some default serialization going on.   And I agree, if it doesn't work with `json.dumps` (I tried it, and it doesn't), then it's busted.   

I have to admit though, it feels wrong to force every notebook that uses pandas with widgets to have to  include serialization code.  I've already got the few lines of code in my notebooks to handle these cases.  Maybe I'll see about incorporating into the declarativewidgets project.

**Comment by minrk** [comment]
Can you show a snippet that worked on 3.2.x and not on 4.x? It's possible that json_clean was silently crushing pandas dataframes to their reprs rather than raising, which would be worse, but I'm pretty sure we used to do that.

**Comment by SylvainCorlay** [comment]
@jtyberg, we have some trait types for common Scipy trait types in bqplot, including some simple serialization.

There is a Jupyter Incubator proposal for Scipy Trait Types here: https://github.com/jupyter-incubator/traittypes. Custom serialization is in the scope of the proposal. The goal for me is to remove the trait types in bqplot and use this proposal as soon as possible.

I have to push the code for serialization but I need traitlets 4.1 to be out to be able to iterate there.

**Comment by jtyberg** [comment]
@minrk In jupyter 3.2.x, I was using version 0.1.1 of the declarativewidgets to send pandas/numpy over the widgets channel indirectly.  I'm not sure `json_clean` was even in the picture, but my widgets code broke when I moved to 4.x and declarativewidgets 0.2.0.

I added a declarativewidgets sample notebook to the same [gist](https://gist.github.com/jtyberg/63fe41605566fc1fe866).  The notebook works in 3.2.x/0.1.1, but breaks in 4.0/0.2.0.
